stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -0,0 +1,121 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+public struct ClassMetadata {
+  var superclassMetadata: UnsafeMutablePointer<ClassMetadata>?
+
+  // There is no way to express the actual calling convention on the heap desroy
+  // function (swiftcc with 'self') currently, so let's use UnsafeRawPointer
+  // and a helper function in C (_swift_runtime_invoke_heap_object_destroy).
+  var destroy: UnsafeRawPointer
+}
+
+public struct HeapObject {
+  var metadata: UnsafeMutablePointer<ClassMetadata>
+  var refcount: Int
+}
+
+@_silgen_name("posix_memalign")
+func posix_memalign(_: UnsafeMutablePointer<UnsafeMutableRawPointer?>, _: UInt, _: UInt) -> CInt
+
+func alignedAlloc(size: UInt, alignment: UInt) -> UnsafeMutableRawPointer? {
+  let alignment = max(alignment, UInt(MemoryLayout<UnsafeRawPointer>.size))
+  var r: UnsafeMutableRawPointer? = nil
+  _ = posix_memalign(&r, alignment, size)
+  return r
+}
+
+@_cdecl("swift_slowAlloc")
+public func swift_slowAlloc(_ size: UInt, _ alignMask: UInt) -> UnsafeMutableRawPointer? {
+  let alignment: UInt
+  if alignMask == UInt.max {
+    alignment = UInt(_swift_MinAllocationAlignment)
+  } else {
+    alignment = alignMask + 1
+  }
+  return alignedAlloc(size: size, alignment: alignment)
+}
+
+@_cdecl("swift_slowDealloc")
+public func swift_slowDealloc(_ ptr: UnsafeMutableRawPointer?, _ size: UInt, _ alignMask: UInt) {
+  free(ptr)
+}
+
+@_silgen_name("swift_allocObject")
+public func swift_allocObject(metadata: UnsafeMutablePointer<ClassMetadata>, requiredSize: UInt, requiredAlignmentMask: UInt) -> UnsafeMutablePointer<HeapObject> {
+  let p = swift_slowAlloc(requiredSize, requiredAlignmentMask)!
+  let object = p.assumingMemoryBound(to: HeapObject.self)
+  object.pointee.metadata = metadata
+  object.pointee.refcount = 1
+  return object
+}
+
+@_silgen_name("swift_deallocClassInstance")
+public func swift_deallocClassInstance(object: UnsafeMutablePointer<HeapObject>, allocatedSize: UInt, allocatedAlignMask: UInt) {
+  free(object)
+}
+
+@_silgen_name("swift_initStackObject")
+public func swift_initStackObject(metadata: UnsafeMutablePointer<ClassMetadata>, object: UnsafeMutablePointer<HeapObject>) -> UnsafeMutablePointer<HeapObject> {
+  object.pointee.metadata = metadata
+  object.pointee.refcount = -1
```

Comment:
Maybe define a static let constant in HeapObject: `static let immortalRefCount = -1`

Reply:
Done.